correctcomputation/checkedc-clang
# Add -output-dir and other file handling improvements

### Draft squash commit message

```
- Add -output-dir option to write updated files to a directory structure
  parallel to the base dir (#347). When -output-dir is used, a source
  file outside the base dir can't be handled because there is no way to
  compute its output path. For consistency, this is now an error even
  when -output-dir is not used.

- Convert all 3C regression tests from -output-postfix to -output-dir to
  avoid leaving temporary files in the clang/test/3C directory (#378).

- Expand "3c -help" documentation. In particular, direct the user to pass
  "--" when they don't want to use a compilation database to avoid
  accidentally using unwanted compiler options and suppress the warning
  if no compilation database is found (#343).

- For consistency, have stdout mode output the main file even if it is
  unchanged (#328).

- Fix bugs in matching of file paths against the base dir (#327).

- Other minor bug fixes: see the pull request description for details.
```

### List of changes and regression testing status

- Fixes #327 (`..`, symlink, `foo.c` is not under `foo/`): all have dedicated tests
- `-output-dir` (fixes #347): Assume this is sufficiently exercised by many of the other tests.
- `-allow-sources-outside-base-dir`: Won’t bother with a test, following the precedent of `-allow-unwritable-changes`.
- New/revised internal functions `getCanonicalFilePath`, `filePathStartsWith`, etc.: Assume these are tested as much as we care via the higher-level features they implement.
- Fixes #328 remaining parts:
  - Multiple source files in stdout mode is now an exit 1 instead of an assertion failure: Not worth testing?
  - For consistency, stdout mode outputs the file even if it is unchanged: sufficiently exercised by other tests
- `-output-dir` + `-output-postfix` is an error: Not worth testing?  [Agreed](https://github.com/correctcomputation/checkedc-clang/pull/412#issuecomment-776860213)
- Error when a source file is outside the base directory: Not worth testing?
- `-base-dir` with trailing slash + `-output-dir` gets canonicalized properly and doesn’t cause files to be written at a wrong path: **update 2021-02-11:** has a dedicated test!
- Nonzero exit when failing to write an output file: Not worth testing?
- Send all debug output to `errs()` instead of `outs()` to avoid corrupting stdout mode: Not worth testing?  [Agreed](https://github.com/correctcomputation/checkedc-clang/pull/412#issuecomment-776860213)
- Revised `3c -help` text, most notably with explanation of `--` (fixes #343) (test N/A)
- Test suite stuff
  - Migrate from `-output-postfix` to `-output-dir` (fixes #378); required adding `-base-dir` in a bunch of places
  - Pass `-base-dir` in many more commands since input files are now required to be under the base dir and the base dir defaults to lit's working directory, which is not `clang/test/3C`
  - Replace `count 0` with `diff` now that stdout mode outputs the file even if it is unchanged
  - `test_updater`: Removed tests from the list that were deleted in #364; mark some tests as no longer producing whitespace diffs
  - All calls to `3c` now use `--`
  - Remove `-fcheckedc-extension` flag from `3c` calls because it got in my way (not from `%clang`; that can be done later in #346)
  - Add `rm -rf %t*` to all tests that mention `%t`.  This is overkill for tests that only use `%t.unused` and the like, but I didn’t think it was worth the trouble to exclude those tests.  This boilerplate is unpleasant but is expected to go away eventually with #355.  I looked into solutions via the lit config and didn’t think it was promising.  See the [LLVM bug report](https://bugs.llvm.org/show_bug.cgi?id=9904).
  - Change the test for an unimplemented `canWrite` constraints case to use a crazy `#include` instead of a typedef, in anticipation of #408.
  - `typedef`, `typedefbounds`: Enable the idempotence tests because they pass and having lines that looked like commands but weren't real `RUN` lines was confusing my scripts.

### TODO

- Migrate all existing regression tests from `-output-postfix` to `-output-dir` (fixes #378)
  - [X] `testgenerator`
  - [X] `processor`
  - [X] `test_updater`
  - [X] Unmanaged tests
  - [X] Delete temporary files before starting the test
  - Managed tests:
    - [X] Revert undesired modifications to stuff other than RUN lines
    - [X] Verify that there were no preexisting diffs in RUN lines
  - [x] Review changes to unmanaged tests
- [X] Fix path canonicalization and comparison bugs (fixes #327) if it isn't too hard
- [X] Review my own code (done except as noted elsewhere)
- [X] Ensure that all changes have regression tests where appropriate (see notes in list of changes)
- [X] Update to latest `main`
- [X] Clean up commit sequence to ease code review
- [X] Draft the final squash commit message
